### PR TITLE
Add first-party Moshi converter.

### PR DESCRIPTION
**Pull request by JakeWharton**

### CODE REVIEW

File: retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    this.mediaType = MediaType.parse("application/json; charset=utf-8");
+  }
+
+  @Override public Object fromBody(ResponseBody body, Type type) throws IOException {
+    BufferedSource source = body.source();
+    try {
+      return moshi.adapter(type).fromJson(source);
+    } finally {
+      try {
+        source.close();
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java]: Should we give source a method like `closeQuietly()` ?!

File: retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  @Override public RequestBody toBody(Object object, Type type) {
+    Buffer buffer = new Buffer();
+    try {
+      moshi.adapter(type).toJson(buffer, object);
+    } catch (IOException e) {
+      throw new AssertionError(e);
+    }
+    return RequestBody.create(mediaType, buffer.snapshot());
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java]: sweeet.

File: retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    ResponseBody body =
+        ResponseBody.create(MediaType.parse("text/plain"), "{\"theName\":\"value\"}");
+    Impl impl = (Impl) converter.fromBody(body, Impl.class);
+    assertEquals("value", impl.getName());
+  }
+
+  private static AbstractCharSequenceAssert<?, String> assertBody(RequestBody body) throws IOException {
+    Buffer buffer = new Buffer();
+    body.writeTo(buffer);
+    return assertThat(buffer.readUtf8());
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java]: I like this pattern.

File: retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  @Test public void serializationTypeUsed() throws IOException {
+    RequestBody body = converter.toBody(new Impl("value"), Example.class);
+    assertBody(body).isEqualTo("{\"name\":\"value\"}");
+  }
+
+  @Test public void deserialization() throws IOException {
+    ResponseBody body =
+        ResponseBody.create(MediaType.parse("text/plain"), "{\"theName\":\"value\"}");
+    Impl impl = (Impl) converter.fromBody(body, Impl.class);
+    assertEquals("value", impl.getName());
```
Comment (swankjesse) [inline retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java]: assertEquals like a caveman

File: retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java (hunk 1 above)
Comment (JakeWharton) [inline retrofit-converters/moshi/src/main/java/retrofit/MoshiConverter.java]: `Source` is an interface and `close()` comes from the `Closable` interface. Even aside from that, I think it'd be out of place to do so.

File: retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java (hunk 4 above)
Comment (JakeWharton) [inline retrofit-converters/moshi/src/test/java/retrofit/MoshiConverterTest.java]: copy/paste from old test. it'll be overhauled by #904

**Pull request merged**